el-patch-template.el

```diff
@@ -0,0 +1,433 @@
+        definition))
+     ((plist-get matched-ptemplate :matched)
+      (error "A template matches multiple forms"))
+     ((and (consp definition)
+           (or
+            (el-patch-template--any-p (car definition)
+                                      ptemplates)
+            (and
+             (cdr definition)
+             (el-patch-template--any-p (cdr definition)
+                                       ptemplates
+                                       (1- matched-forms-count)))))
+      (error "A form which matches a template has subforms which\
+ match other templates"))
+     (t
+      ;; The old resolution of the template uniquely matches the definition
+      ;; Here we first mark the template as being matched then
+      ;; do the actual resolution
+      (plist-put matched-ptemplate :matched t)
+      (let ((resolution
+             (el-patch-template--process definition
+                                         (list
+                                          (plist-get matched-ptemplate
+                                                     :template)))))
+        (cons (caar resolution)
+              (el-patch-template--apply-impl (cdr resolution)
+                                             ptemplates)))))))
+
+(defun el-patch-template--apply (definition templates)
+  "Looks for all templates in TEMPLATES inside definition and
+process them. Returns the processed forms after splicing in
+el-patch directives."
+  (let ((ptemplates (mapcar
+                     (lambda (template)
+                       (list :template template
+                             :old (el-patch--resolve template nil)
+                             :matched nil))
+                     templates)))
+    (prog1 (el-patch-template--apply-impl definition ptemplates)
+      (cl-dolist (ptemplate ptemplates)
+        (unless (plist-get ptemplate :matched)
+          (error "At least one template did not match any form"))))))
+
+(defmacro el-patch-template (keyword-name &rest templates)
+  "KEYWORD-NAME is cons whose car is a type which can be any type
+from `el-patch-deftype-alist' and the cdr is the name of the form
+to be patched. Looks for all forms which match a template in
+TEMPLATES and processes them by matching all `...' forms against
+the souce code of the form, while keeping the `el-patch-*'
+directives. Returns an `el-patch-*' definition."
+  `(let* ((qkeyword-name (quote ,keyword-name))
+          (definition (el-patch--locate
+                       (car (el-patch--resolve qkeyword-name
+                                               nil))))
+          patched)
+     (unless definition
+       (error "Cannot find definition for `%s'"
+              (cadr qkeyword-name)))
+     (setq patched (el-patch-template--apply definition
+                                             ',templates))
+     ;; NOTE: Unfortunately `el-patch-deftype-alist' doesn't save the
+     ;; macro name so we have to assume that it is `el-patch-*'
```

Comment:
I wonder if `el-patch-deftype-alist' should save the macro-name if it is provided to `el-patch-deftype`.